[GEOS-11346] Add a configurable Content-Security-Policy header (#7514)
geoserver/geoserver: doc/en/user/source/security/csp.rst

  understand what the policy does.
* The :guilabel:`Enabled` checkbox will enable/disable the policy.

.. note::
    After saving a policy, make sure to save/apply the configuration.

Configuring Rules
-----------------

Each rule contains a filter to match against user requests and the CSP directives to add to the
header value for matching requests. Rules will be checked in order against incoming requests and
only the first matching rule in each policy will be applied. If no rule in a policy matches the
request, then no directives will be added to the CSP header from that policy. If the matching
rule has no directives defined, then preceding rules will be checked until the first rule is
found that has directives and no directives will be added if no such rule exists.

The button for adding rules can be found at the top of the :guilabel:`Rule List` table and a rule
can be edited by clicking on its name in the table or removed by clicking on the remove icon at the
end of the rule's row in the table. Rule positions can also be changed by using the up/down arrows
or by dragging and dropping the rule's row in the table.

.. figure:: images/csp-rule.png

>    CSP Rule Configuration

Reviewer: We need a new example as the WFS Test Servlet is not included any more.

Author: Done

  match the key_regex have a URL-decoded value that match the value_regex. The value regex will be
  tested against an empty string if no query parameters matched the key regex.
  Example: ``PARAM((?i)^service$,(?i)^wms$)``
* PROP(key,value_regex): Returns true if the value for the property key matches the regex. The key
  is case-sensitive and must contain must contain the string ``GeoServer``, ``GeoTools``, or
  ``GeoWebCache`` anywhere in the key (case-insensitive). The regex will be tested against an empty
  string if the property is not set. This is primarily intended for the default configuration and
  may not be useful to administrators.
  Example: ``PROP(GEOSERVER_CONSOLE_DISABLED,(?i)^(?!true$).*$)``

.. note::
    The ``(?i)`` at the beginning of the regular expression will use case-insensitive matching and
    enclosing the pattern inside of the ``^$`` characters will match the entire string. See the
    `Regular Expressions Tutorial <https://docs.oracle.com/javase/tutorial/essential/regex/>`_ for
    more information about how to use Java regular expressions.

Leaving the filter blank will cause this rule to match all requests and should only be used on the

> last rule in a policy since any additional rules would never be checked.

Reviewer: That seems really hard to understand; if blank is aways used then it does not need to be shown (or maybe indicated another way).

.. _security_csp:

Content Security Policy
=======================

> The :guilabel:`Content Security Policy` page controls the checks that are performed on GeoServer
> requests that GeoServer will use to set an appropriate Content-Security-Policy (CSP) HTTP
> response header to mitigate cross-site scripting and clickjacking attacks.

Reviewer 2: ```suggestion
The :guilabel:`Content Security Policy` page controls how GeoServer prepares the Content-Security-Policy (CSP) HTTP response headers, used to mitigate cross-site scripting and clickjacking attacks.
```

.. note::

>     The settings on this page do not affect anything if the CSP header has been disabled using the

Reviewer 2: ```suggestion
    The settings on this page do not affect anything if the CSP headers have been disabled using the
```

    ``geoserver.csp.shouldSetPolicy`` property or if the ``geoserver.csp.policy`` property is used
    to set a fixed header value.

.. note::
    Starting with version 9, Wicket will set its own CSP header with a very strict policy that will

>     overwrite what is set by GeoServer although there is a workaround in place to merge directives

Reviewer 2: ```suggestion
    overwrite what is set by GeoServer, although there is a workaround in place to merge directives
```

>     overwrite what is set by GeoServer although there is a workaround in place to merge directives

Reviewer 2: This is a useful note for me (I was about to ask indeed), but not for an end user, that does not know which version of Wicket we are using. Until the Wicket update branch is merged, we should inform the user about the behavior, based on the current Wicket version.

Suggestion: update this comment, and once it's merged, propose a documentation change in the Wicket upgrade branch.

Author: Other than referencing specific Wicket versions, this comment would still be useful to users after the Wicket update. It lets them know that disabling GeoServer's CSP will not disable the CSP on Wicket pages and also explains that the actual CSP header for GeoServer's Wicket pages will be a combination of Wicket's and GeoServer's CSP.

Author: The documentation no longer refers to specific Wicket version. The comment may need to be updated if a Wicket upgrade does not happen in the 2.26 release.

Author: I removed the comment from the documentation for now and added a note in CSPDefaultConfiguration.java that the default configuration and documentation need to be updated after the Wicket 9 upgrade.

    set by GeoServer that are either not supported or not set by default by Wicket (particularly
    ``form-action`` and ``frame-ancestors``). Wicket's CSP is separate from GeoServer's
    functionality and will otherwise not be affected by GeoServer's configuration.

Default Configuration
---------------------

The default CSP configuration is intended to support many GeoServer use cases and allow users to
securely run GeoServer without having to modify the configuration. It may be updated in future
releases to fix bugs, support new features or enhance security.

The default header value for most GeoServer requests will be (without line breaks):

::
    base-uri 'self'; form-action 'self'; default-src 'none'; child-src 'self'; connect-src 'self';
    font-src 'self'; img-src 'self' data:; style-src 'self' 'unsafe-inline'; script-src 'self';,

>     frame-ancestors 'self';

Reviewer 2: Not sure what you wanted to get here, but the rendering is weird (happens in the other examples in the pages as well):

![image](https://github.com/geoserver/geoserver/assets/325433/7dd018f4-4ae4-4c48-bbb6-0b57db019a41)

Also, is this comma intended? ``script-src 'self';,``

Author: The rendering is fixed now.

change the header value at the top of this page to:

::
    base-uri 'self'; form-action 'self'; default-src 'none'; child-src 'self'; connect-src 'self';
    font-src 'self'; img-src 'self' data:; style-src 'self' 'unsafe-inline'; script-src 'self';,
    frame-ancestors 'self' https://geoserver.org;

Configuring Policies
--------------------

Each policy contains the rules and directives to set a single CSP header value. When there are CSP
directives from multiple policies, the directives will be concatenated into a single line using
commas rather than setting multiple Content-Security-Policy headers. If a value for a specific
directive is defined in multiple policies, web browsers will use the strictest value set for that
directive.

The button for adding policies can be found at the top of the :guilabel:`Policy List` table and a
policy can be edited by clicking on its name in the table or removed by clicking on the remove icon
at the end of the policy's row in the table. Policy positions can also be changed by using the
up/down arrows or by dragging and dropping the rule's row in the table.

> .. figure:: images/csp-policy.png

Reviewer 2: The screenshot is scaled down and becomes unreadable.
